### PR TITLE
Assign auto populated columns on Active Record object creation 

### PR DESCRIPTION
**Pull request by nvasilevski**
This PR extends Active Record creation logic to allow for a database
auto-populated attributes to be assigned on object creation.

Given a `Post` model represented by the following schema:
```ruby
create_table :posts, id: false do |t|
  t.integer :sequential_number, auto_increment: true
  t.string :title, primary_key: true
  t.string :ruby_on_rails, default: -> { "concat('R', 'o', 'R')" }
end
```
where `title` is being used as a primary key, the table has an
integer `sequential_number` column populated by a sequence and
`ruby_on_rails` column has a default function - creation of
`Post` records should populate the `sequential_number` and
`ruby_on_rails` attributes:

```ruby
new_post = Post.create(title: 'My first post')
new_post.sequential_number # => 1
new_post.ruby_on_rails # => 'RoR'
```

_At this moment MySQL and SQLite adapters are limited to only one
column being populated and the column must be the `auto_increment`
while PostgreSQL adapter supports any number of auto-populated
columns through `RETURNING` statement._

### Implementation details

Overall solution is trying to be generic and extensible to allow for additions while still maintains (for MySQL and SQLite) the fact that currently Rails only populates a single column on creation.

- The most important change I'd like to bring the most attention to is the method signature change for both `exec_insert` and `insert` public methods. We are adding a new `returning` keyword which accepts either a single value or an array of columns to control the return value of the method and define what column values to be returned after row insertion. Currently only supported by Postgresql using `RETURNING` statement. The behavior of the `returning` argument is defined as following: `nil` is the default value and doesn't change the behavior of the methods - they continue to return a single primary key or `LAST_INSERT_ID()` value. If an array of column names is passed - an array is returned back. However non-`nil` value for the argument is only supported by postgresql adapter at the moment. 
- `ConnectionAdapters::Column` gets a new adapter-abstract attribute called `auto_incremented_by_db?` which indicates whether a given column is being auto-incremented by the database regardless of the way it's being incremented. For mysql we alias it to `auto_increment` and for postgresql we alias it to `serial`. 
- `ConnectionAdapters::Column` also gets a new `auto_populated?` method which allows us to identify columns auto-populated by the database regardless of the way how the column is being populated. Currently only checks for `auto_incremented_by_db` or a default function
- `SQLite3::Column` has a new `rowid` property which is an implicit way to have an integer column being autoincremented
- `ModelSchema._returning_columns_for_insert` has been added to gather a list of columns values for which we want to get back after insertion. Adapter defines the selection rule. postgresql will select any auto-populated column while mysql and sqlite will select only `auto_incremented_by_db?` which by implication limits the selection to a single column.

### Long term plans

The feature is flexible enough to be extended to support not just postgresql adapter but any other database that is capable of returning column values after creation. 
The closest candidates are:
- [MariaDB starting with 10.5.0](https://mariadb.com/kb/en/insertreturning/)
- [SQLite since version 3.35.0 (2021-03-12).](https://www.sqlite.org/lang_returning.html)

We'll have to teach adapters how to build a `RETURNING` statement in `sql_for_insert`
https://github.com/rails/rails/blob/aa162cefe19b81daa5d7f8c2b9ee4fe2564e73fc/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L632-L633
and then just extend the `return_value_after_insert?` method to ask for more than just an auto-increment being returned after an insert. 
https://github.com/rails/rails/blob/aa162cefe19b81daa5d7f8c2b9ee4fe2564e73fc/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L284-L285

Databases that don't support `RETURNING` statement or an alternative will require an additional `SELECT` query to be performed to fetch values for every auto-populated columns. An additional query may not be desirable for every application so the behavior should be configurable which brings more complexity and one of the reasons why we are not trying to tackle it all at once.

### CODE REVIEW

**Comment by eileencodes** [comment]
Can you take a look at the test failures when you're back on Thursday?

**Comment by matthewd** [comment]
> fetches only one column as insert signature is only capable of returning a single value

I suspect this might be the best moment to change that: we can vary the return value by whether the new parameter is supplied.

That does mean our caller is going to need to know how to figure out which return strategy the adapter supports, but I think that's already true: prioritising the PK is sensible, but in the case of MySQL, it's not true that we can get the new value for one column of our choosing [e.g. if the PK is _not_ auto-populated] -- we will always get the value of the autoincrement column, whatever it is, and otherwise nothing. 99% of the time that's equivalent, but for when it's not, I think we do need to know about the distinction.

(That is, by my reading, the current implementation would be wrong on MySQL when faced with a PK that defaults to `RAND() * 1000` and a second integer column that's `AUTO_INCREMENT`. I'm not sure it's practical for us to make that _not_ broken, but we at least shouldn't shove values into incorrect attributes.)

**Comment by nvasilevski** [comment]
>I suspect this might be the best moment to change that: we can vary the return value by whether the new parameter is supplied.

Right, for some reason I thought we can't do it without a deprecation cycle but we can if we only change the signature in case if the new `returning_columns` argument is used. Though I don't feel completely comfortable about the duality of the method return value, it reminds me of the "boolean flag argument" smell so I'm thinking of maybe defining new `exec_insert_with_returning` and `insert_with_returning` methods instead. It may result in some duplication though. I'll try both "separate methods" and "returning_columns" argument approaches to see which one makes more sense and causes less troubles but let me know if you already have preference. 

My only concern here that regardless of implementation the multi-return values solution will only work with PostgreSQL unless we want to add a special case for non-postgresql databases which will make an additional query to fetch values for all auto-populated columns. Though it may be too much for a single PR 

>the current implementation would be wrong on MySQL when faced with a PK that defaults to RAND() * 1000 and a second integer column that's AUTO_INCREMENT.

That's correct. Not an excuse but just a clarification that current implementation on `main` won't be able to handle such scenario either as it explicitly assumes that whichever value is returned by `LAST_INSERT_ID()` is the primary key value
https://github.com/rails/rails/blob/7d1c4dd7be1b51e2dcbb3614b68dc98c2926d28c/activerecord/lib/active_record/persistence.rb#L1242

Though I agree that since we are making changes we have a great opportunity to fix it as well. Because it is impossible to have more than one `AUTO_INCREMENT` column in MySQL (I'm relying on the error `ERROR 1075 (42000) at line 3: Incorrect table definition; there can be only one auto column and it must be defined as a key`) the strategy for MySQL (and most likely for SQLite)  will be implemented as "populate **the** (only) `auto_increment?` column while for postgresql it will be "populate everything" 


I'll work on extending the return values of insert methods (either through an argument or using a new method) along with a strategy, will post an update when its ready

**Comment by nvasilevski** [comment]
I've made some changes to the PR. I'll be updating the description and verifying the change against a real application on CI but here is the summary: 
`exec_insert_with_returning` and `insert_with_returning` are two new `:nodoc:` methods which now store bodies of `exec_insert` and `insert` methods.  `exec_insert` and `insert` methods are now special cases of `_with_returning` methods that keep their signatures and return values. `_with_returning` name might not be the prettiest but we are not rushing with making these methods public so we can always rename it before exposing it. Having `_with_returning` methods being capable of passing multiple columns down to the adapter and return multiple column values in result allowed postgresql to fully benefit from it and allow for multiple auto-populated columns to be assigned right after the creation. 
The default behavior is to only "return" `auto_incremented_by_db?` which allows for non-postgresql adapters to assign the `AUTO_INCREMENT` column regardless of its relation to the `primary_key`. I'm primarily relying on the fact that MySQL doesn't allow for more than one column to be `AUTO_INCREMENT` and SQLite being even more strict and forcing the `AUTO_INCREMENT` to be a single-column primary key.

**Review by matthewd — APPROVED** [review]
Looks great!

I think it's worth some follow-up to make sure adapters that don't know how to do full-featured RETURNING will instead raise if they're asked to return a column(s) they can't... but no need to squeeze that into this PR.

**Comment by nvasilevski** [comment]
>I think it's worth some follow-up to make sure adapters that don't know how to do full-featured RETURNING will instead raise if they're asked to return a column(s) they can't

Thanks, that's being done right now in parallel and a PR will be ready soon!

**Comment by jrochkind** [comment]
This is very welcome, I've been hoping for behavior like this for some years, as in a languishing PR at https://github.com/rails/rails/pull/39968.  Thanks very much for this!

It looks like some of the API is still using the terminology "auto-incrementing", which may not be appropriate, as per example at top here, it can be used for things that aren't just "auto-incrementing", but "auto-populating", eg the example at top with `default: -> { "concat('R', 'o', 'R')" }`

The code and comments in the diff sometimes uses the terminology "auto-incrementing" but other times "auto-populating" -- are there places that could/should be made more consistent?

It also is not clear to me from looking at the diff/docs if I need to configure something to get ActiveRecord, at the high level, to automatically pick up postgres values from columns for which I've defined a database-level default, that I want to auto-populate on insert with postgres `RETURNING` -- or if this will somehow happen completely automatically detected. (Or neither, something else?)  Can anyone confirm/explain here, for others that may be looking for docs on this exciting new feature too?  (And I wonder if additional info in ActiveRecord docs/guides somewhere, or CHANGELOG would be helpful?)

Again, thank you for this!

**Comment by rafaelfranca** [comment]
> If i have a postgres non-pk column where the database supplies default values on insert, how can I get ActiveRecord to auto-populate that on insert?

You just don't set in on the insert and it will work.

**Comment by nvasilevski** [comment]
>The code and comments in the diff sometimes uses the terminology "auto-incrementing" but other times "auto-populating" -- are there places that could/should be made more consistent?

So in the code we do have two new concepts - `auto_incremented_by_db?` and `auto_populated?` but we are not using those interchangeably and each place intentionally uses the one it needs. `auto_incremented_by_db?` is a method that helps us abstracting column auto increment property from specific adapters since there are multiple adapter-specific properties like `rowid` or `auto_increment` or `serial` that all imply the column being auto-incremented.  `auto_populated?` is a higher-level concept that just tells that a column will be populated by the database regardless of an actual way of setting the value. 

So while in the code I'd say that we are using these concepts correctly I agree that in descriptions and the changelog it may not be as clear. This primarily caused by the need to account for all databases as the fix works slightly differently depending on the database used. In a few words:

**PostgreSQL** - any number of "db-auto-populated" attributes will be assigned after record creation
**MySQL & SQLite3** - unfortunately most applications won't see any difference, at least not yet. However there is a corner case which will be affected by this change. It is mainly specific for tables with a composite primary key, where an `auto_increment` column does not solely represent the primary key. 

So it wasn't easy to properly describe this change as for postgresql it means full support of auto-populated columns being assigned on creation while for other databases it has much more limited impact

I'll try to see if there is any suitable place in guides or docs to add more clarifications. I don't think it belongs to `create` or `save` method docs since from some point of view this behavior is expected by default and the docs for persistence methods may become unnecessarily verbose and duplicated.

Basically we need to find a balance between making sure that PostgreSQL users aware of the full potential of the feature but we also need to avoid misleading MySQL & SQLite3 users 

>if I need to configure something

As Rafael pointed out, it will just start working out of the box. The list of the columns to be auto-assigned after creation is automatically determined but if some columns are being missed feel free to report it as a bug! Hopefully it should be easily extensible by tweaking the `auto_populated?` definition.
